File: src/hardware.py
"""
Consolidated Hardware Module
=============================
Simplified hardware abstractions for BlueBuzzah v2.

This module consolidates all hardware interfaces into a single file:
- Board configuration and pin assignments
- I2C multiplexer control
- DRV2605 haptic controller
- Battery voltage monitoring
- NeoPixel LED control

Module: hardware
Version: 2.0.0
"""

import time
from core.types import ActuatorType, BatteryStatus
from core.constants import (
    DRV2605_DEFAULT_ADDR,
    I2C_MULTIPLEXER_ADDR,
    MAX_AMPLITUDE,
    MIN_FREQUENCY_HZ,
    MAX_FREQUENCY_HZ,
    CRITICAL_VOLTAGE,
    LOW_VOLTAGE,
)


# ============================================================================
# Board Configuration
# ============================================================================

class BoardConfig:
    """
    Board-specific configuration for nRF52840.

    Attributes:
        board_id: Board identifier string
        neopixel_pin: Pin for NeoPixel LED
        battery_sense_pin: Pin for battery voltage ADC
        i2c: I2C bus instance
        tca9548a_address: I2C multiplexer address
    """

    def __init__(self):
        """Initialize board configuration with pin assignments."""
        import board
        import busio

        self.board_id = getattr(board, "board_id", "nrf52840")

        # Pin assignments
        self.neopixel_pin = board.NEOPIXEL
        self.battery_sense_pin = board.VOLTAGE_MONITOR if hasattr(board, "VOLTAGE_MONITOR") else board.A6

        # I2C bus configuration
        self.i2c = busio.I2C(board.SCL, board.SDA, frequency=400000)

        # I2C multiplexer address
        self.tca9548a_address = I2C_MULTIPLEXER_ADDR


# ============================================================================
# I2C Multiplexer
# ============================================================================

class I2CMultiplexer:
    """
    Simple TCA9548A I2C multiplexer controller.

    The multiplexer enables communication with multiple I2C devices
    (haptic drivers) sharing the same address on different channels.

    Attributes:
        i2c: I2C bus instance
        address: Multiplexer I2C address
        active_channel: Currently selected channel
    """

    def __init__(self, i2c, address= I2C_MULTIPLEXER_ADDR):
        """
        Initialize I2C multiplexer.

        Args:
            i2c: I2C bus instance
            address: Multiplexer address (default: 0x70)
        """
        self.i2c = i2c
        self.address = address
        self.active_channel= None

        # Verify multiplexer is present
        try:
            self.deselect_all()
        except Exception as e:
            raise RuntimeError(f"Multiplexer not found at 0x{address:02X}: {e}")

    def select_channel(self, channel):
        """
        Select multiplexer channel (0-7).

        Args:
            channel: Channel number

        Raises:
            ValueError: If channel out of range
        """
        if not (0 <= channel <= 7):
            raise ValueError(f"Channel {channel} out of range (0-7)")

        try:
            channel_mask = 1 << channel
            # Acquire I2C lock for CircuitPython
            while not self.i2c.try_lock():
                pass
            try:
                self.i2c.writeto(self.address, bytes([channel_mask]))
                self.active_channel = channel
            finally:
                self.i2c.unlock()
        except Exception as e:
            raise RuntimeError(f"Failed to select channel {channel}: {e}")

    def deselect_all(self):
        """Disable all multiplexer channels."""
        try:
            # Acquire I2C lock for CircuitPython
            while not self.i2c.try_lock():
                pass
            try:
                self.i2c.writeto(self.address, bytes([0x00]))
                self.active_channel = None
            finally:
                self.i2c.unlock()
        except Exception as e:
            raise RuntimeError(f"Failed to deselect channels: {e}")


# ============================================================================
# Haptic Controller (DRV2605)
# ============================================================================

class DRV2605Controller:
    """
    DRV2605 haptic driver controller for up to 5 actuators.

    Controls vibrotactile actuators (LRA or ERM) through I2C multiplexer,
    using real-time playback (RTP) mode for precise amplitude control.

    Attributes:
        multiplexer: I2C multiplexer instance
        actuator_type: LRA or ERM
        i2c_addr: DRV2605 I2C address
        active_fingers: Dict tracking active state per finger
        frequencies: Dict storing frequency per finger
    """

    # DRV2605 Registers
    REG_MODE = 0x01
    REG_RTP_INPUT = 0x02
    REG_FEEDBACK_CTRL = 0x1A
    REG_CONTROL1 = 0x1B
    REG_CONTROL3 = 0x1D

    # Mode values
    MODE_REALTIME_PLAYBACK = 0x05

    def __init__(
        self,
        multiplexer,
        actuator_type= ActuatorType.LRA,
        frequency_hz= 175,
        i2c_addr= DRV2605_DEFAULT_ADDR,
    ):
        """
        Initialize DRV2605 controller.

        Args:
            multiplexer: I2C multiplexer instance
            actuator_type: LRA or ERM
            frequency_hz: Default resonant frequency (150-250 Hz)
            i2c_addr: DRV2605 I2C address (default: 0x5A)
        """
        self.multiplexer = multiplexer
        self.actuator_type = actuator_type
        self.i2c_addr = i2c_addr
        self.default_frequency = frequency_hz

        self.active_fingers= {}
        self.frequencies= {}

    def initialize_finger(self, finger):
        """
        Initialize DRV2605 for specific finger.

        Args:
            finger: Finger index (0-4)
        """
        if not (0 <= finger < 5):
            raise ValueError(f"Finger {finger} out of range (0-4)")

        self.multiplexer.select_channel(finger)

        try:
            # Exit standby
            self._write_register(self.REG_MODE, 0x00)

            # DRV2605 datasheet section 7.4.2:
            # "Wait 5ms after exiting standby before writing to registers"
            time.sleep(0.005)

            # Configure for actuator type
            if self.actuator_type == ActuatorType.LRA:
                self._write_register(self.REG_FEEDBACK_CTRL, 0x80)  # LRA mode
                self._write_register(self.REG_CONTROL3, 0x01)  # LRA mode bit
            else:
                self._write_register(self.REG_FEEDBACK_CTRL, 0x00)  # ERM mode
                self._write_register(self.REG_CONTROL3, 0x00)  # ERM mode

            # Set RTP mode
            self._write_register(self.REG_MODE, self.MODE_REALTIME_PLAYBACK)

            # Set frequency if LRA
            if self.actuator_type == ActuatorType.LRA:
                self.set_frequency(finger, self.default_frequency)

            self.active_fingers[finger] = False

        finally:
            self.multiplexer.deselect_all()

    def activate(self, finger, amplitude):
        """
        Activate haptic motor at specified amplitude.

        Args:
            finger: Finger index (0-4)
            amplitude: Vibration amplitude (0-100%)
        """
        if not (0 <= finger < 5):
            raise ValueError(f"Finger {finger} out of range (0-4)")
        if not (0 <= amplitude <= MAX_AMPLITUDE):
            raise ValueError(f"Amplitude {amplitude} out of range (0-{MAX_AMPLITUDE})")

        # Convert 0-100% to 0-127
        rtp_value = int((amplitude / 100.0) * 127)

        self.multiplexer.select_channel(finger)

        try:
            self._write_register(self.REG_RTP_INPUT, rtp_value)
            self.active_fingers[finger] = True
        finally:
            self.multiplexer.deselect_all()

    def deactivate(self, finger):
        """
        Deactivate haptic motor.

        Args:
            finger: Finger index (0-4)
        """
        if not (0 <= finger < 5):
            raise ValueError(f"Finger {finger} out of range (0-4)")

        self.multiplexer.select_channel(finger)

        try:
            self._write_register(self.REG_RTP_INPUT, 0x00)
            self.active_fingers[finger] = False
        finally:
            self.multiplexer.deselect_all()

    def set_frequency(self, finger, frequency):
        """
        Set resonant frequency for LRA actuator.

        Args:
            finger: Finger index (0-4)
            frequency: Resonant frequency (150-250 Hz)
        """
        if not (0 <= finger < 5):
            raise ValueError(f"Finger {finger} out of range (0-4)")
        if not (MIN_FREQUENCY_HZ <= frequency <= MAX_FREQUENCY_HZ):
            raise ValueError(f"Frequency {frequency} Hz out of range")

        self.frequencies[finger] = frequency

        # Calculate drive time
        drive_time = int(5000 / frequency) & 0x1F

        self.multiplexer.select_channel(finger)

        try:
            self._write_register(self.REG_CONTROL1, drive_time)
        finally:
            self.multiplexer.deselect_all()

    def is_active(self, finger):
        """
        Check if finger's motor is active.

        Args:
            finger: Finger index (0-4)

        Returns:
            True if active, False otherwise
        """
        return self.active_fingers.get(finger, False)

    def stop_all(self):
        """Stop all haptic motors immediately."""
        for finger in range(5):
            if self.active_fingers.get(finger, False):
                try:
                    self.multiplexer.select_channel(finger)
                    self._write_register(self.REG_RTP_INPUT, 0x00)
                    self.active_fingers[finger] = False
                except Exception:
                    pass  # Best effort
                finally:
                    self.multiplexer.deselect_all()

    def emergency_stop(self):
        """Emergency stop all motors (synchronous version)."""
        self.stop_all()

    def _write_register(self, register, value):
        """
        Write to DRV2605 register.

        Args:
            register: Register address
            value: Value to write (0-255)
        """
        try:
            i2c = self.multiplexer.i2c
            buffer = bytearray([register, value])
            # Acquire I2C lock for CircuitPython
            while not i2c.try_lock():
                pass
            try:
                i2c.writeto(self.i2c_addr, buffer)
            finally:
                i2c.unlock()
        except Exception as e:
            raise RuntimeError(f"Failed to write register 0x{register:02X}: {e}")


# ============================================================================
# Battery Monitor
# ============================================================================

class BatteryMonitor:
    """
    Battery voltage monitoring for LiPo batteries.

    Monitors battery voltage through ADC pin with voltage divider,
    calculates percentage, and checks threshold levels.

    Attributes:
        pin: AnalogIn pin for voltage reading
        divider_ratio: Voltage divider ratio
        sample_count: Number of samples to average
        warning_voltage: Low battery warning threshold
        critical_voltage: Critical battery shutdown threshold
    """

    ADC_REF_VOLTAGE = 3.3
    ADC_MAX_VALUE = 65535

    # LiPo voltage curve (voltage -> percentage)
    VOLTAGE_CURVE = [
        (4.20, 100), (4.15, 95), (4.11, 90), (4.08, 85), (4.02, 80),
        (3.98, 75), (3.95, 70), (3.91, 65), (3.87, 60), (3.85, 55),
        (3.84, 50), (3.82, 45), (3.80, 40), (3.79, 35), (3.77, 30),
        (3.75, 25), (3.73, 20), (3.71, 15), (3.69, 10), (3.61, 5), (3.27, 0),
    ]

    def __init__(
        self,
        adc_pin,
        divider_ratio= 2.0,
        sample_count= 10,
        warning_voltage= LOW_VOLTAGE,
        critical_voltage= CRITICAL_VOLTAGE,
    ):
        """
        Initialize battery monitor.

        Args:
            adc_pin: Pin object for ADC voltage reading
            divider_ratio: Voltage divider ratio
            sample_count: Number of ADC samples to average
            warning_voltage: Low battery warning threshold
            critical_voltage: Critical battery threshold
        """
        # Create AnalogIn from pin for ADC reading
        try:
            import analogio
            self.pin = analogio.AnalogIn(adc_pin)
        except ImportError:
            # Mock for testing on non-CircuitPython platforms
            self.pin = adc_pin

        self.divider_ratio = divider_ratio
        self.sample_count = sample_count
        self.warning_voltage = warning_voltage
        self.critical_voltage = critical_voltage
        self.last_status= None

    def read_voltage(self):
        """
        Read battery voltage with averaging.

        Returns:
            Battery voltage in volts
        """
        total = 0
        for _ in range(self.sample_count):
            total += self.pin.value
            time.sleep(0.001)

        average = total / self.sample_count
        adc_voltage = (average / self.ADC_MAX_VALUE) * self.ADC_REF_VOLTAGE
        battery_voltage = adc_voltage * self.divider_ratio

        return battery_voltage

    def get_percentage(self, voltage= None):
        """
        Calculate battery percentage from voltage.

        Args:
            voltage: Battery voltage (if None, reads from ADC)

        Returns:
            Battery percentage (0-100)
        """
        if voltage is None:
            voltage = self.read_voltage()

        if voltage >= self.VOLTAGE_CURVE[0][0]:
            return 100
        if voltage <= self.VOLTAGE_CURVE[-1][0]:
            return 0

        # Linear interpolation
        for i in range(len(self.VOLTAGE_CURVE) - 1):
            v_high, pct_high = self.VOLTAGE_CURVE[i]
            v_low, pct_low = self.VOLTAGE_CURVE[i + 1]

            if v_low <= voltage <= v_high:
                ratio = (voltage - v_low) / (v_high - v_low)
                percentage = pct_low + ratio * (pct_high - pct_low)
                return int(percentage)

        return 0

    def get_status(self, voltage= None):
        """
        Get comprehensive battery status.

        Args:
            voltage: Battery voltage (if None, reads from ADC)

        Returns:
            BatteryStatus dataclass
        """
        if voltage is None:
            voltage = self.read_voltage()

        percentage = self.get_percentage(voltage)

        # Determine status
        if voltage < self.critical_voltage:
            status_str = "CRITICAL"
            is_low = True
            is_critical = True
        elif voltage < self.warning_voltage:
            status_str = "LOW"
            is_low = True
            is_critical = False
        else:
            status_str = "OK"
            is_low = False
            is_critical = False

        battery_status = BatteryStatus(
            voltage=voltage,
            percentage=percentage,
            status=status_str,
            is_low=is_low,
            is_critical=is_critical,
        )

        self.last_status = battery_status
        return battery_status

    def is_low(self, voltage= None):
        """Check if battery is below warning threshold."""
        if voltage is None:
            voltage = self.read_voltage()
        return voltage < self.warning_voltage

    def is_critical(self, voltage= None):
        """Check if battery is below critical threshold."""
        if voltage is None:
            voltage = self.read_voltage()
        return voltage < self.critical_voltage


# ============================================================================
# NeoPixel LED
# ============================================================================

class LEDPin:
    """
    Simple NeoPixel LED wrapper.

    Attributes:
        pin: Board pin for NeoPixel
        pixel: NeoPixel instance
    """

    def __init__(self, pin):
        """
        Initialize LED pin.

        Args:
            pin: Board pin object
        """
        import neopixel

        self.pin = pin
        self.pixel = neopixel.NeoPixel(pin, 1, brightness=0.1, auto_write=True)

    def set_color(self, r, g, b):
        """
        Set LED color.

        Args:
            r: Red value (0-255)
            g: Green value (0-255)
            b: Blue value (0-255)
        """
        self.pixel[0] = (r, g, b)

    def off(self):
        """Turn LED off."""
        self.pixel[0] = (0, 0, 0)

File: src/hardware.cpp
/**
 * @file hardware.cpp
 * @brief BlueBuzzah hardware abstraction layer - Implementations
 * @version 2.0.0
 * @platform Adafruit Feather nRF52840 Express
 */

#include "hardware.h"

// =============================================================================
// BATTERY MONITOR - Static Data
// =============================================================================

// LiPo discharge curve: {voltage, percentage}
// 21 calibration points for accurate interpolation
const float BatteryMonitor::VOLTAGE_CURVE[][2] = {
    {4.20f, 100}, {4.15f, 95}, {4.11f, 90}, {4.08f, 85}, {4.02f, 80},
    {3.98f, 75},  {3.95f, 70}, {3.91f, 65}, {3.87f, 60}, {3.85f, 55},
    {3.84f, 50},  {3.82f, 45}, {3.80f, 40}, {3.79f, 35}, {3.77f, 30},
    {3.75f, 25},  {3.73f, 20}, {3.71f, 15}, {3.69f, 10}, {3.61f, 5},
    {3.27f, 0}
};

const uint8_t BatteryMonitor::VOLTAGE_CURVE_SIZE = sizeof(VOLTAGE_CURVE) / sizeof(VOLTAGE_CURVE[0]);

// =============================================================================
// HAPTIC CONTROLLER - Implementation
// =============================================================================

HapticController::HapticController() : _tca(TCA9548A_ADDRESS), _initialized(false) {
    for (uint8_t i = 0; i < MAX_ACTUATORS; i++) {
        _fingerActive[i] = false;
        _fingerEnabled[i] = false;
    }
}

bool HapticController::begin() {
    Serial.println(F("[INFO] Initializing haptic controller..."));

    // Initialize I2C at 400kHz
    Wire.begin();
    Wire.setClock(I2C_FREQUENCY);

    // Initialize TCA9548A multiplexer
    _tca.begin(Wire);
    _tca.closeAll();

    Serial.printf("[INFO] TCA9548A multiplexer initialized at 0x%02X\n", TCA9548A_ADDRESS);

    // Initialize each finger's DRV2605
    uint8_t successCount = 0;
    for (uint8_t finger = 0; finger < MAX_ACTUATORS; finger++) {
        if (initializeFinger(finger)) {
            successCount++;
            Serial.printf("[INFO] DRV2605 finger %d initialized\n", finger);
        } else {
            Serial.printf("[ERROR] DRV2605 finger %d initialization failed\n", finger);
        }
    }

    _initialized = (successCount > 0);

    if (_initialized) {
        Serial.printf("[INFO] Haptic controller ready: %d/%d fingers enabled\n",
                      successCount, MAX_ACTUATORS);
    } else {
        Serial.println(F("[CRITICAL] No haptic drivers initialized!"));
    }

    return _initialized;
}

bool HapticController::initializeFinger(uint8_t finger) {
    if (finger >= MAX_ACTUATORS) {
        return false;
    }

    // Retry logic for reliable initialization
    for (uint8_t attempt = 0; attempt < I2C_RETRY_COUNT; attempt++) {
        if (attempt > 0) {
            Serial.printf("[WARN] Retrying finger %d init (attempt %d/%d)\n",
                          finger, attempt + 1, I2C_RETRY_COUNT);
            delay(I2C_RETRY_DELAY_MS);
        }

        // Select multiplexer channel
        if (!selectChannel(finger)) {
            continue;
        }

        // Initialize DRV2605 on this channel
        if (!_drv[finger].begin()) {
            closeChannels();
            continue;
        }

        // Apply initialization delay for I2C stabilization
        delay(I2C_INIT_DELAY_MS);

        // Configure for LRA + RTP mode
        configureDRV2605(_drv[finger]);

        // SAFETY: Immediately stop motor after init - DRV2605 retains RTP value
        // across MCU resets, so motor may be buzzing from pre-power-off state
        _drv[finger].setRealtimeValue(0);

        closeChannels();

        // Mark as enabled
        _fingerEnabled[finger] = true;
        _fingerActive[finger] = false;

        return true;
    }

    // All retries failed
    _fingerEnabled[finger] = false;
    return false;
}

void HapticController::configureDRV2605(Adafruit_DRV2605& drv) {
    // 1. Configure for LRA (Linear Resonant Actuator)
    drv.useLRA();

    // 2. Enable open-loop mode (v1 requirement for proper LRA operation)
    // Register 0x1D (CONTROL3): Set bits 5 (N_PWM_ANALOG) and 0 (LRA_OPEN_LOOP)
    uint8_t control3 = drv.readRegister8(0x1D);
    drv.writeRegister8(0x1D, control3 | 0x21);

    // 3. Set peak voltage to 2.50V (v1 default: ACTUATOR_VOLTAGE = 2.50)
    // Register 0x17 (OD_CLAMP): voltage = value * 0.02122V
    // 2.50V / 0.02122 ≈ 118
    drv.writeRegister8(0x17, 118);

    // 4. Set driving frequency to 250Hz (v1 default)
    // Register 0x20 (OL_LRA_PERIOD): period = 1 / (freq * 0.00009849)
    // 250Hz: 1 / (250 * 0.00009849) ≈ 40
    drv.writeRegister8(0x20, 40);

    // 5. Set Real-Time Playback (RTP) mode
    drv.setMode(DRV2605_MODE_REALTIME);

    // 6. Initialize RTP value to 0 (motor off)
    drv.setRealtimeValue(0);
}

bool HapticController::selectChannel(uint8_t finger) {
    if (finger >= MAX_ACTUATORS) {
        return false;
    }

    _tca.openChannel(finger);
    return true;
}

void HapticController::closeChannels() {
    _tca.closeAll();
}

uint8_t HapticController::amplitudeToRTP(uint8_t amplitude) const {
    // Clamp amplitude to valid range
    if (amplitude > MAX_AMPLITUDE) {
        amplitude = MAX_AMPLITUDE;
    }

    // Convert 0-100% to 0-127
    return (uint8_t)((amplitude * DRV2605_MAX_RTP) / MAX_AMPLITUDE);
}

Result HapticController::activate(uint8_t finger, uint8_t amplitude) {
    // Validate finger index
    if (finger >= MAX_ACTUATORS) {
        return Result::ERROR_INVALID_PARAM;
    }

    // Check if finger is enabled
    if (!_fingerEnabled[finger]) {
        return Result::ERROR_DISABLED;
    }

    // Validate amplitude
    if (amplitude > MAX_AMPLITUDE) {
        amplitude = MAX_AMPLITUDE;
    }

    // Select channel
    if (!selectChannel(finger)) {
        return Result::ERROR_HARDWARE;
    }

    // Set RTP value
    uint8_t rtpValue = amplitudeToRTP(amplitude);
    _drv[finger].setRealtimeValue(rtpValue);

    closeChannels();

    // Update state
    _fingerActive[finger] = (amplitude > 0);

    return Result::OK;
}

Result HapticController::deactivate(uint8_t finger) {
    // Validate finger index
    if (finger >= MAX_ACTUATORS) {
        return Result::ERROR_INVALID_PARAM;
    }

    // Check if finger is enabled
    if (!_fingerEnabled[finger]) {
        return Result::ERROR_DISABLED;
    }

    // Select channel
    if (!selectChannel(finger)) {
        return Result::ERROR_HARDWARE;
    }

    // Set RTP value to 0
    _drv[finger].setRealtimeValue(0);

    closeChannels();

    // Update state
    _fingerActive[finger] = false;

    return Result::OK;
}

void HapticController::stopAll() {
    for (uint8_t finger = 0; finger < MAX_ACTUATORS; finger++) {
        if (_fingerEnabled[finger] && _fingerActive[finger]) {
            deactivate(finger);
        }
    }
}

void HapticController::emergencyStop() {
    // Stop all motors regardless of tracked state
    for (uint8_t finger = 0; finger < MAX_ACTUATORS; finger++) {
        if (_fingerEnabled[finger]) {
            if (selectChannel(finger)) {
                _drv[finger].setRealtimeValue(0);
            }
        }
    }
    closeChannels();

    // Clear all active states
    for (uint8_t i = 0; i < MAX_ACTUATORS; i++) {
        _fingerActive[i] = false;
    }
}

bool HapticController::isActive(uint8_t finger) const {
    if (finger >= MAX_ACTUATORS) {
        return false;
    }
    return _fingerActive[finger];
}

bool HapticController::isEnabled(uint8_t finger) const {
    if (finger >= MAX_ACTUATORS) {
        return false;
    }
    return _fingerEnabled[finger];
}

Result HapticController::setFrequency(uint8_t finger, uint16_t frequencyHz) {
    // Validate finger index
    if (finger >= MAX_ACTUATORS) {
        return Result::ERROR_INVALID_PARAM;
    }

    // Validate frequency range
    if (frequencyHz < MIN_FREQUENCY_HZ || frequencyHz > MAX_FREQUENCY_HZ) {
        return Result::ERROR_INVALID_PARAM;
    }

    // Check if finger is enabled
    if (!_fingerEnabled[finger]) {
        return Result::ERROR_DISABLED;
    }

    // Select channel
    if (!selectChannel(finger)) {
        return Result::ERROR_HARDWARE;
    }

    // Calculate drive time for LRA frequency
    // Formula from DRV2605 datasheet
    uint8_t driveTime = (uint8_t)((5000 / frequencyHz) & 0x1F);

    // Write to CONTROL1 register (0x1B)
    _drv[finger].writeRegister8(DRV2605_REG_CONTROL1, driveTime);

    closeChannels();

    return Result::OK;
}

uint8_t HapticController::getEnabledCount() const {
    uint8_t count = 0;
    for (uint8_t i = 0; i < MAX_ACTUATORS; i++) {
        if (_fingerEnabled[i]) {
            count++;
        }
    }
    return count;
}

// =============================================================================
// BATTERY MONITOR - Implementation
// =============================================================================

BatteryMonitor::BatteryMonitor() : _initialized(false) {
}

bool BatteryMonitor::begin() {
    // Configure ADC resolution
    analogReadResolution(ADC_RESOLUTION_BITS);

    // Take a test reading to verify ADC works
    uint32_t testReading = analogRead(BATTERY_PIN);

    // Any non-zero reading indicates the ADC is working
    _initialized = (testReading > 0);

    Serial.println(F("[INFO] Battery monitor initialized"));

    return _initialized;
}

float BatteryMonitor::readVoltage() {
    if (!_initialized) {
        return 0.0f;
    }

    // Take multiple samples and average for stability
    uint32_t total = 0;
    for (uint8_t i = 0; i < BATTERY_SAMPLE_COUNT; i++) {
        total += analogRead(BATTERY_PIN);
        delay(1);
    }

    float average = (float)total / BATTERY_SAMPLE_COUNT;

    // Convert ADC reading to voltage
    // Formula: (adc_value / max_value) * reference_voltage * divider_ratio
    float voltage = (average / (float)ADC_MAX_VALUE) * ADC_REFERENCE_VOLTAGE * BATTERY_VOLTAGE_DIVIDER;

    return voltage;
}

uint8_t BatteryMonitor::getPercentage(float voltage) {
    if (voltage < 0) {
        voltage = readVoltage();
    }

    return interpolatePercentage(voltage);
}

uint8_t BatteryMonitor::interpolatePercentage(float voltage) const {
    // Handle edge cases
    if (voltage >= VOLTAGE_CURVE[0][0]) {
        return 100;
    }
    if (voltage <= VOLTAGE_CURVE[VOLTAGE_CURVE_SIZE - 1][0]) {
        return 0;
    }

    // Find the two points to interpolate between
    for (uint8_t i = 0; i < VOLTAGE_CURVE_SIZE - 1; i++) {
        float v1 = VOLTAGE_CURVE[i][0];
        float v2 = VOLTAGE_CURVE[i + 1][0];

        if (voltage <= v1 && voltage > v2) {
            // Linear interpolation
            float p1 = VOLTAGE_CURVE[i][1];
            float p2 = VOLTAGE_CURVE[i + 1][1];

            float ratio = (voltage - v2) / (v1 - v2);
            float percentage = p2 + ratio * (p1 - p2);

            return (uint8_t)percentage;
        }
    }

    // Fallback (should not reach here)
    return 0;
}

BatteryStatus BatteryMonitor::getStatus() {
    BatteryStatus status;

    status.voltage = readVoltage();
    status.percentage = interpolatePercentage(status.voltage);
    status.isLow = (status.voltage < BATTERY_LOW_VOLTAGE);
    status.isCritical = (status.voltage < BATTERY_CRITICAL_VOLTAGE);

    return status;
}

bool BatteryMonitor::isLow(float voltage) {
    if (voltage < 0) {
        voltage = readVoltage();
    }
    return voltage < BATTERY_LOW_VOLTAGE;
}

bool BatteryMonitor::isCritical(float voltage) {
    if (voltage < 0) {
        voltage = readVoltage();
    }
    return voltage < BATTERY_CRITICAL_VOLTAGE;
}

// =============================================================================
// LED CONTROLLER - Implementation
// =============================================================================

LEDController::LEDController()
    : _pixel(NEOPIXEL_COUNT, NEOPIXEL_PIN, NEO_GRB + NEO_KHZ800),
      _baseColor(0, 0, 0),
      _displayColor(0, 0, 0),
      _pattern(LEDPattern::OFF),
      _initialized(false),
      _patternStartTime(0),
      _blinkState(false),
      _lastBlinkToggle(0) {
}

bool LEDController::begin() {
    _pixel.begin();
    _pixel.setBrightness(LED_BRIGHTNESS);
    _pixel.clear();
    _pixel.show();

    _initialized = true;
    _patternStartTime = millis();

    Serial.println(F("[INFO] LED controller initialized"));

    return _initialized;
}

void LEDController::update() {
    if (!_initialized) {
        return;
    }

    uint32_t now = millis();

    switch (_pattern) {
        case LEDPattern::SOLID:
            // No animation needed - color already applied
            break;

        case LEDPattern::BREATHE_SLOW: {
            float brightness = calculateBreatheBrightness(LED_BREATHE_SLOW_MS);
            RGBColor modulated(
                (uint8_t)(_baseColor.r * brightness),
                (uint8_t)(_baseColor.g * brightness),
                (uint8_t)(_baseColor.b * brightness)
            );
            if (modulated != _displayColor) {
                applyColor(modulated);
            }
            break;
        }

        case LEDPattern::PULSE_SLOW: {
            float brightness = calculateBreatheBrightness(LED_PULSE_SLOW_MS);
            RGBColor modulated(
                (uint8_t)(_baseColor.r * brightness),
                (uint8_t)(_baseColor.g * brightness),
                (uint8_t)(_baseColor.b * brightness)
            );
            if (modulated != _displayColor) {
                applyColor(modulated);
            }
            break;
        }

        case LEDPattern::BLINK_FAST: {
            uint32_t interval = _blinkState ? LED_BLINK_FAST_ON_MS : LED_BLINK_FAST_OFF_MS;
            if (now - _lastBlinkToggle >= interval) {
                _lastBlinkToggle = now;
                _blinkState = !_blinkState;
                applyColor(_blinkState ? _baseColor : Colors::OFF);
            }
            break;
        }

        case LEDPattern::BLINK_SLOW: {
            uint32_t interval = _blinkState ? LED_BLINK_SLOW_ON_MS : LED_BLINK_SLOW_OFF_MS;
            if (now - _lastBlinkToggle >= interval) {
                _lastBlinkToggle = now;
                _blinkState = !_blinkState;
                applyColor(_blinkState ? _baseColor : Colors::OFF);
            }
            break;
        }

        case LEDPattern::BLINK_URGENT: {
            uint32_t interval = _blinkState ? LED_BLINK_URGENT_ON_MS : LED_BLINK_URGENT_OFF_MS;
            if (now - _lastBlinkToggle >= interval) {
                _lastBlinkToggle = now;
                _blinkState = !_blinkState;
                applyColor(_blinkState ? _baseColor : Colors::OFF);
            }
            break;
        }

        case LEDPattern::BLINK_CONNECT: {
            uint32_t interval = _blinkState ? LED_BLINK_CONNECT_ON_MS : LED_BLINK_CONNECT_OFF_MS;
            if (now - _lastBlinkToggle >= interval) {
                _lastBlinkToggle = now;
                _blinkState = !_blinkState;
                applyColor(_blinkState ? _baseColor : Colors::OFF);
            }
            break;
        }

        case LEDPattern::OFF:
            // LED is off, nothing to update
            break;
    }
}

void LEDController::setPattern(const RGBColor& color, LEDPattern pattern) {
    if (!_initialized) {
        return;
    }

    _baseColor = color;
    _pattern = pattern;
    _patternStartTime = millis();
    _lastBlinkToggle = millis();
    _blinkState = true;  // Start blink patterns in ON state

    // Immediately apply the initial state
    switch (pattern) {
        case LEDPattern::SOLID:
            applyColor(color);
            break;

        case LEDPattern::BREATHE_SLOW:
        case LEDPattern::PULSE_SLOW:
            // Start at full brightness, update() will animate
            applyColor(color);
            break;

        case LEDPattern::BLINK_FAST:
        case LEDPattern::BLINK_SLOW:
        case LEDPattern::BLINK_URGENT:
        case LEDPattern::BLINK_CONNECT:
            // Start in ON state
            applyColor(color);
            break;

        case LEDPattern::OFF:
            applyColor(Colors::OFF);
            break;
    }
}

void LEDController::setColor(uint8_t r, uint8_t g, uint8_t b) {
    setPattern(RGBColor(r, g, b), LEDPattern::SOLID);
}

void LEDController::setColor(const RGBColor& color) {
    setPattern(color, LEDPattern::SOLID);
}

void LEDController::off() {
    setPattern(Colors::OFF, LEDPattern::OFF);
}

void LEDController::setBrightness(uint8_t brightness) {
    if (!_initialized) {
        return;
    }

    // Cap brightness to prevent external code from exceeding max
    uint8_t cappedBrightness = (brightness > LED_BRIGHTNESS) ? LED_BRIGHTNESS : brightness;
    _pixel.setBrightness(cappedBrightness);
    _pixel.show();
}

RGBColor LEDController::getColor() const {
    return _baseColor;
}

LEDPattern LEDController::getPattern() const {
    return _pattern;
}

void LEDController::applyColor(const RGBColor& color) {
    _displayColor = color;
    _pixel.setPixelColor(0, _pixel.Color(color.r, color.g, color.b));
    _pixel.show();
}

float LEDController::calculateBreatheBrightness(uint32_t cycleMs) const {
    // Calculate position in cycle (0.0 to 1.0)
    uint32_t elapsed = millis() - _patternStartTime;
    float position = (float)(elapsed % cycleMs) / (float)cycleMs;

    // Use sine wave for smooth breathing effect
    // sin goes from -1 to 1, we want 0 to 1
    // Position 0 = brightness 0.5, position 0.25 = brightness 1.0
    // position 0.5 = brightness 0.5, position 0.75 = brightness 0.0
    float brightness = (sin(position * 2.0f * PI - PI / 2.0f) + 1.0f) / 2.0f;

    // Clamp to ensure we don't go below a minimum visibility threshold
    const float MIN_BRIGHTNESS = 0.1f;
    return MIN_BRIGHTNESS + brightness * (1.0f - MIN_BRIGHTNESS);
}
